tests/test_ports.py

```diff
@@ -0,0 +1,121 @@
+import kfactory as kf
+import pytest
+from enum import IntEnum
+
+
+def _l(layer: int, datatype: int):
+    return kf.library.layer(layer, datatype)
+
+
+class LAYER(IntEnum):
+    SI = _l(1, 0)
+    M1 = _l(2, 0)
+
+    def __str__(self):
+        return self.name
+
+
+@kf.autocell
+def waveguide(width: int, length: int, layer: int) -> kf.KCell:
+    c = kf.KCell()
+
+    c.shapes(LAYER.SI).insert(kf.kdb.Box(0, -width // 2, length, width // 2))
+
+    c.create_port(
+        name="o1", trans=kf.kdb.Trans(2, False, 0, 0), width=width, layer=layer
+    )
+    c.create_port(
+        name="o2", trans=kf.kdb.Trans(0, False, length, 0), width=width, layer=layer
+    )
+    return c
+
+
+@pytest.fixture()
+def wg():
+    return waveguide(1000, 20000, LAYER.SI)
+
+
+@pytest.fixture()
+@kf.autocell
+def wg_floating_off_grid():
+    c = kf.KCell()
+    dbu = c.library.dbu
+
+    p1 = kf.kcell.DPort(
+        width=10 + dbu / 2,
+        name="o1",
+        trans=kf.kdb.DTrans(2, False, dbu / 2, 0),
+        layer=LAYER.SI,
+    )
+    p2 = kf.kcell.DPort(
+        width=10 + dbu / 2,
+        name="o2",
+        trans=kf.kdb.DTrans(0, False, 20 + dbu, 0),
+        layer=LAYER.SI,
+    )
+    c.shapes(LAYER.SI).insert(kf.kdb.DBox(p1.x, -p1.width / 2, p2.x, p1.width / 2))
+    c.add_port(p1)
+    c.add_port(p2)
+
+    c.draw_ports()
+    return c
```

Comment:
@joamatab this is an example of why I am of the opinion that complex and/or floating point ports will introduce unwanted sideeffects even ignoring that the connection functions are a lot more complicated.